### PR TITLE
DAOS-623 build: Add AddressSanitizer build mode

### PR DESCRIPTION
**Pull request by jolivier23**
BUILD_TYPE=asan and using gcc, this sets -fsanitize=address
so we can do some runtime checks of stack overflows and
leaks
Fix one leak in the timestamp test

Signed-off-by: Jeff Olivier <jeffrey.v.olivier@intel.com>

### CODE REVIEW

File: SConstruct [hunk 1]
```diff
@@ -114,6 +114,25 @@ def is_platform_arm():
         return True
     return False
 
+def check_asan_helper(context):
```
Comment (ashleypittman) [inline SConstruct:117]: I can't see what this is doing that's asan specific?

File: SConstruct [hunk 2]
```diff
@@ -132,6 +151,17 @@ def set_defaults(env):
             env.AppendUnique(CCFLAGS=['-Og'])
         else:
             env.AppendUnique(CCFLAGS=['-O0'])
+    elif env.get('BUILD_TYPE') == 'asan':
+        flags = ['-fsanitize=address', '-O1', '-fno-omit-frame-pointer']
+        env.AppendUnique(CCFLAGS=flags)
+        env.AppendUnique(CFLAGS=flags)
+        env.AppendUnique(CXXFLAGS=flags)
+        env.AppendUnique(LINKFLAGS=['-fsanitize=address', '-O1'])
```
Comment (ashleypittman) [inline SConstruct:159]: Oh, is it just linking with this library?  That would make sense.

File: src/vos/tests/vts_ts.c [hunk 3]
```diff
@@ -761,7 +761,10 @@ static int
 ts_test_fini(void **state)
 {
 	struct ts_test_arg	*ts_arg = *state;
+	struct vos_ts_table	*ts_table;
 
+	ts_table = vos_ts_table_get();
+	vos_ts_table_free(&ts_table);
```
Comment (ashleypittman) [inline src/vos/tests/vts_ts.c:767]: Is this a current memory leak?  Is so I'd like to try this with my leak checker before we fix it.

File: utils/sl/prereq_tools/base.py [hunk 4]
```diff
@@ -819,9 +819,14 @@ def has_source(self, env, *comps, **kw):
     def _setup_build_type(self):
         """set build type"""
         self.add_opts(EnumVariable('BUILD_TYPE', "Set the build type",
-                                   'dev', ['dev', 'debug', 'release'],
+                                   'dev', ['dev', 'debug', 'release', 'asan'],
```
Comment (ashleypittman) [inline utils/sl/prereq_tools/base.py:822]: Is there a reason this wouldn't go in the debug or dev builds?

I wonder if we shouldn't have a list of build options, and have dev,debug,release map to a set of options.

Comment (jolivier23) [reply utils/sl/prereq_tools/base.py:822]: Not a bad idea to have a map

File: SConstruct (hunk 1 above)
Comment (jolivier23) [reply SConstruct:117]: I should probably move the options into this routine

File: SConstruct [hunk 5]
```diff
@@ -114,6 +114,25 @@ def is_platform_arm():
...
+    context.Message("Checking for AddressSanitizer... ")
+    ret = context.TryLink("int main() { return 0; }", ".c")
+    context.Result(ret)
+    return ret
+
+def check_asan(env):
+    """Check for address sanitizer support"""
+    result = False
+    cenv = env.Clone()
+    config = Configure(cenv, custom_tests={"CheckASan" : check_asan_helper})
```
Comment (frostedcmos) [inline SConstruct:128]: why AS vs As in 'CheckASan' ?

File: SConstruct [hunk 6]
```diff
@@ -115,6 +115,25 @@ def is_platform_arm():
         return True
     return False
 
+def check_asan_helper(context):
+    """configure callback to check for address sanitizer"""
+    context.Message("Checking for AddressSanitizer... ")
+    ret = context.TryLink("int main() { return 0; }", ".c")
+    context.Result(ret)
+    return ret
+
```
Comment (jolivier23) [inline SConstruct:124]:
```suggestion
print("Hello World")
```

**Pull request closed without merging**